h/storage.py

```diff
@@ -147,6 +147,10 @@ def create_annotation(request, data, group_service):
     request.db.add(annotation)
     request.db.flush()
 
+    request.find_service(name="search_index")._queue.add_by_id(
+        annotation.id, tag="storage.create_annotation", schedule_in=30
+    )
```

Comment:
I don't think this is the right place to do it. Currently:

 * In `h.views.api.annotations.create()` we call `storage.create_annotation`
 * Then we call `_publish_annotation_event` which creates `AnnotationEvent`
 * This is picked up by `h.subscribers.annotation_sync()`
 * This calls `h.services.search_index.service.handle_annotation_event()`
 * This is where we do primary indexing (or async fallback if that fails)

I think this last point is where any addition 3rd layer back up should take place.

Reply:
This can't be done in that event subscriber because this has to happen within the request's original DB transaction: the atomic transactionality (so it definitely can't create/update an annotation in the DB without also adding the job) is a crucial aspect.

Reply:
@jon-betts Do you happen to know what the upper time limit is on how long the "synchronous" annotation indexing can take, if it uses the maximum number of Celery retries? We should set the `schedule_in=30` to be comfortably longer than that

Reply:
The synchronous indexing doesn't involve rabbit at all, so there shouldn't be any retries. It should be the total time we would take to obtain an ES handle from a pool I assume (which may have retries) and perform the update.

I don't know how long that is, but the max timeouts I've seen in NewRelic for ES calls are about 20s, so 30s would be safe.

Reply:
It [does send the task to Rabbit it Elasticsearch fails](https://github.com/hypothesis/h/blob/5f207cd5ca428ae915bbe08196417ff39d52a429/h/services/search_index/service.py#L130), right? I think in the worse case scenario where Elasticsearch fails and the maximum amount of Celery retries get used, we don't want the job queue to kick in until _after_ all of those retries

Reply:
Clarification from slack, but the "synchronous" here we mean the initial attempt in process + the delayed celery attempts.

The synchronous indexing doesn't involve rabbit at all, so there shouldn't be any retries. It should be the total time we would take to obtain an ES handle from a pool I assume (which may have retries) and perform the update.

I don't know how long that is, but the max timeouts I've seen in NewRelic for ES calls are about 20s, so 30s would be safe (assuming this includes the time to get a handle). If the handle has a separate timeout, it could be longer.

The celery task has a similar situation obtaining a handle from the pool, but the two outcomes here are pretty much:

* It gets one right away
* It totally fails (presumably at the max timeout)

The timeout to get a handle is less than 2s by `RETRY_POLICY_QUICK`) but assuming that works and we then go onto fail repeatedly while retrying, the second scenario would probably be longer and dominate.

We use exponential back off and jitter up to 5 times. So without jitter I think it's 1 + 2 + 4 + 8 + 16 (32), but jitter treats all of these values as a maximum, so an average of 16s max 32s, ignoring time to do anything.

So in summary in the worst case

 * However long it takes to get an ES handle
 * + however long it takes ES to timeout
 * + How long it takes to timeout connecting to Rabbit (<2s)
 * + 32s (retry delays)
 * + any time spent in task doing things and failing

So I'd say more minutes than seconds.

Reply:
I think the Elasticsearch timeout is actually 5 seconds. But I think to be on the safe side we could `schedule_in=180` (three mins), that seems fine to me